### ISSUE TITLE
[Support]: USB edge TPU not detected

### ISSUE THREAD

**Issue by enp6s0** [blakeblackshear/frigate#2178]
### Describe the problem you are having

I have a Coral USB edge TPU passed through to a VM running in ESXi (as a USB 3.0 device). I could see the device both inside the Frigate container and the host VM, but Frigate is crashing on start complaining that the TPU couldn't be detected:

```
# lsusb -tv
/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/4p, 5000M
    ID 1d6b:0003 Linux Foundation 3.0 root hub
    |__ Port 1: Dev 3, If 0, Class=Application Specific Interface, Driver=, 5000M
        ID 1a6e:089a Global Unichip Corp. 
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/4p, 480M
    ID 1d6b:0002 Linux Foundation 2.0 root hub
    |__ Port 1: Dev 2, If 0, Class=Human Interface Device, Driver=usbhid, 12M
        ID 0e0f:0003 VMware, Inc. Virtual Mouse
```

```
# ls -la /dev | grep coral
lrwxrwxrwx   1 root       root          15 Nov  3 19:28 coral -> bus/usb/002/003

# ls -la /dev/bus/usb/002/003
crwxrwxrwx 1 root plugdev 189, 130 Nov  3 19:28 /dev/bus/usb/002/003

```

This is what I have for `docker-compose.yml`:
```
version: "3.9"
services:
  frigate:
    container_name: frigate
    privileged: true
    restart: always
    image: blakeblackshear/frigate:stable-amd64
    shm_size: "1024mb"
    devices:
      - /dev/bus/usb:/dev/bus/usb # coral usb
      - /dev/dri/renderD128 # Intel HWaccel for video decoding
    volumes:
      # - /dev/bus/usb:/dev/bus/usb # coral usb?
      - /etc/localtime:/etc/localtime:ro
      - /frigate/config.yml:/config/config.yml:ro
      - /var/footage:/media/frigate
      - type: tmpfs
        target: /tmp/cache
        tmpfs:
          size: 1000000000
    ports:
      - "5000:5000"
      - "1935:1935" # RTMP feeds
    environment:
      FRIGATE_RTSP_PASSWORD: "secret"

```

What I've tried (none have worked so far, TPU is still not detected):

- Moving `/dev/bus/usb` into `volumes` (as suggested by another comment on another issue)
- Setting permission for the Coral device itself to `0777` (through udev)
- Physically changing the USB port the edge TPU is connected to (normally it's connected through a USB 3.0 hub, I tried a direct connection)

Would appreciate some help and/or debugging tips!

### Version

0.9.4-26ae608

### Frigate config file

```yaml
mqtt:
  host: <mqtt server ip>

detectors:
  coral:
    type: edgetpu
    device: usb

cameras:
  01_test_camera:
    ffmpeg:
      inputs:
        - path: <camera rtsp path>
          roles:
            - detect
            - rtmp
    detect:
      width: 1920
      height: 1080
```

### Relevant log output

```shell
frigate  | [2021-11-03 19:20:45] ws4py                          INFO    : Using epoll
frigate  | [2021-11-03 19:20:44] detector.coral                 INFO    : Starting detection process: 212
frigate  | [2021-11-03 19:20:44] frigate.edgetpu                INFO    : Attempting to load TPU as usb
frigate  | [2021-11-03 19:20:54] frigate.edgetpu                ERROR   : No EdgeTPU was detected. If you do not have a Coral device yet, you must configure CPU detectors.
frigate  | Process detector:coral:
frigate  | Traceback (most recent call last):
frigate  |   File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 160, in load_delegate
frigate  |     delegate = Delegate(library, options)
frigate  |   File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 119, in __init__
frigate  |     raise ValueError(capture.message)
frigate  | ValueError
frigate  | 
frigate  | During handling of the above exception, another exception occurred:
frigate  | 
frigate  | Traceback (most recent call last):
frigate  |   File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
frigate  |     self.run()
frigate  |   File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
frigate  |     self._target(*self._args, **self._kwargs)
frigate  |   File "/opt/frigate/frigate/edgetpu.py", line 156, in run_detector
frigate  |     object_detector = LocalObjectDetector(
frigate  |   File "/opt/frigate/frigate/edgetpu.py", line 64, in __init__
frigate  |     edge_tpu_delegate = load_delegate("libedgetpu.so.1.0", device_config)
frigate  |   File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 162, in load_delegate
frigate  |     raise ValueError('Failed to load delegate from {}\n{}'.format(
frigate  | ValueError: Failed to load delegate from libedgetpu.so.1.0
```

### FFprobe output from your camera

```shell
Likely not relevant?
```

### Operating system

Debian

### Install method

Docker Compose

### Coral version

USB

### Network connection

Wired

### Camera make and model

N/A

**Comment by ozett** [comment]
check this?
![image](https://user-images.githubusercontent.com/5937811/140279724-7d39e12e-713e-44f2-bc53-648f3ddea70c.png)

**Comment by ozett** [comment]
same plattform? synology? 
mabe same solution? add a powered usb-hub?

**Comment by blakeblackshear** [comment]
@enp6s0 I think I have heard some others talk about needing to pass the entire usb hub to the VM rather than a single USB device. I am not very familiar with ESX, so I'm not sure how to do it. Did you try searching for older issues with "ESX"?

**Comment by ozett** [comment]
take this shortcut? -> https://github.com/blakeblackshear/frigate/issues/305

**Comment by bigramonk** [comment]
> same plattform? synology? mabe same solution? add a powered usb-hub?

SOLVED - for some unknown reason, the Coral USB port had gone back to its default setting in Proxmox (host=1a6e:089a) and not anymore in a **pass-through mode**, which is required for Frigate to detect it within Home Assistant.

Thanks!

**Comment by enp6s0** [comment]
@blakeblackshear it might be an issue with USB passthrough. After rebooting the ESXi host, the Coral was properly recognized, but it only appears to work only once (after I stopped Frigate, the USB device just completely disappeared and whatever I do I couldn't get it to show up again).

```
/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/4p, 5000M
    ID 1d6b:0003 Linux Foundation 3.0 root hub
    |__ Port 1: Dev 2, If 0, Class=Vendor Specific Class, Driver=usbfs, 5000M
        ID 18d1:9302 Google Inc. 
```

Previous issues on ESXi seems to suggest that it is possible to use the TPU in a VM, but so far it seems that success is limited to PCIe TPUs (or passthrough of the entire USB controller, which I couldn't do due to hardware limitations). The article that @ozett  (and the documentation) linked to used a PCIe TPU.

I'll play around with it a bit more to see if this could be made to work (found [this](https://github.com/google-coral/edgetpu-platforms/issues/4) while searching around, but it's not exactly ESXi). Might be worth considering putting in a note in the docs that USB TPUs are flaky under ESXi...

**Comment by ozett** [comment]
> Might be worth considering putting in a note in the docs that USB TPUs are flaky under ESXi...

maybe a problem of version or hardware-version, 
so that would be good, if mentioned in the docs, to provide ESX-Version and maybe Hardware-details.
so one can later see for wich version a problem was identified..

**Comment by xawen** [comment]
I spent a few days trying to get USB passthrough to work into an ESXi VM.  What I found is that it won't work consistently.  If you do manage to get it recognized, a reboot will probably break it again.  

From what I can tell, the reason for this is that the Coral is an odd device.  When you first plug it in, it will show up as a "Global Unichip" device with one device ID.  After it initializes, it will change to a "Google USB Device" with a different ID.  The Google USB Device is the one you want to use, but it's not visible to ESXi until it's initialized...but that can't happen until the VM is attached to it.

This short version is that, from my testing, the only way to make it work is to pass the entire USB controller into the VM.  I did this by buying a cheap PCIe USB3 card for my host and dedicating the whole card to my Home Assistant VM.  It's been rock solid under that configuration.

**Comment by brentco** [comment]
> I spent a few days trying to get USB passthrough to work into an ESXi VM. What I found is that it won't work consistently. If you do manage to get it recognized, a reboot will probably break it again.
> 
> From what I can tell, the reason for this is that the Coral is an odd device. When you first plug it in, it will show up as a "Global Unichip" device with one device ID. After it initializes, it will change to a "Google USB Device" with a different ID. The Google USB Device is the one you want to use, but it's not visible to ESXi until it's initialized...but that can't happen until the VM is attached to it.
> 
> This short version is that, from my testing, the only way to make it work is to pass the entire USB controller into the VM. I did this by buying a cheap PCIe USB3 card for my host and dedicating the whole card to my Home Assistant VM. It's been rock solid under that configuration.

Thanks for the unichip hint. I was passing through the TPU to a VM with frigate running on it. It stopped working for some reason. I found indeed that first there was a unichip device and in fact it remained like that until I passed it on to frigate. I am guessing frigate initialized the TPU which then caused the USB device to change (different IDs). Passing through both (by device id) fixed the issue for me. I found both of them by using `lsusb -v`.

**Comment by rvalitov** [comment]
In my case the problem was in the stock cable that came with the Coral TPU. It seems it works only for power delivery but not for data transfer. When I changed the cable the Coral TPU started to be detected in Linux, Windows, Proxmox, etc.

**Comment by rolfvreijdenberger** [comment]
in my case (proxmox host, home assistant VM with usb passthrough and frigate in container addon) I had the issue with usb passthrough, which is described in this blog: https://stijn.tintel.eu/blog/2025/01/22/load-google-coral-usb-firmware-udev/ (fix is present)
